From 5d7d53fd4200258e635a52012b7b4f8207df3a9b Mon Sep 17 00:00:00 2001
From: Chunjie Zhu <chunjie.zhu@cloud.com>
Date: Tue, 22 Aug 2023 09:37:10 +0000
Subject: [PATCH] CA-72973: hack to strip temp dirs from paths

---
 python/automake.mk | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/python/automake.mk b/python/automake.mk
index 767512f17..fe7ea4cd8 100644
--- a/python/automake.mk
+++ b/python/automake.mk
@@ -70,16 +70,17 @@ FLAKE8_PYFILES += \
 	python/ovs/dirs.py.template
 
 nobase_pkgdata_DATA = $(ovs_pyfiles) $(ovstest_pyfiles)
+base = $(shell echo $$RPM_BUILD_ROOT)
 ovs-install-data-local:
 	$(MKDIR_P) python/ovs
 	sed \
 		-e '/^##/d' \
-		-e 's,[@]pkgdatadir[@],$(pkgdatadir),g' \
-		-e 's,[@]RUNDIR[@],$(RUNDIR),g' \
+		-e 's,[@]pkgdatadir[@],$(subst $(base),,$(pkgdatadir)),g' \
+		-e 's,[@]RUNDIR[@],$(subst $(base),,$(RUNDIR)),g' \
 		-e 's,[@]LOGDIR[@],$(LOGDIR),g' \
-		-e 's,[@]bindir[@],$(bindir),g' \
-		-e 's,[@]sysconfdir[@],$(sysconfdir),g' \
-		-e 's,[@]DBDIR[@],$(DBDIR),g' \
+		-e 's,[@]bindir[@],$(subst $(base),,$(bindir)),g' \
+		-e 's,[@]sysconfdir[@],$(subst $(base),,$(sysconfdir)),g' \
+		-e 's,[@]DBDIR[@],$(subst $(base),,$(DBDIR)),g' \
 		< $(srcdir)/python/ovs/dirs.py.template \
 		> python/ovs/dirs.py.tmp
 	$(MKDIR_P) $(DESTDIR)$(pkgdatadir)/python/ovs
-- 
2.34.1

