diff --git a/ofproto/bond.c b/ofproto/bond.c
index 7dbbb80..8a37aad 100644
--- a/ofproto/bond.c
+++ b/ofproto/bond.c
@@ -1664,6 +1664,8 @@ bond_slave_lookup(struct bond *bond, const void *slave_)
 static void
 bond_enable_slave(struct bond_slave *slave, bool enable)
 {
+    struct bond *bond = slave->bond;
+
     slave->delay_expires = LLONG_MAX;
     if (enable != slave->enabled) {
         slave->bond->bond_revalidate = true;
@@ -1673,6 +1675,7 @@ bond_enable_slave(struct bond_slave *slave, bool enable)
         if (enable) {
             list_insert(&slave->bond->enabled_slaves, &slave->list_node);
         } else {
+            bond->send_learning_packets = true;
             list_remove(&slave->list_node);
         }
         ovs_mutex_unlock(&slave->bond->mutex);
