diff --git a/xenserver/etc_xapi.d_plugins_openvswitch-cfg-update b/xenserver/etc_xapi.d_plugins_openvswitch-cfg-update
index ff6044c..2319af4 100755
--- a/xenserver/etc_xapi.d_plugins_openvswitch-cfg-update
+++ b/xenserver/etc_xapi.d_plugins_openvswitch-cfg-update
@@ -30,7 +30,7 @@ import re
 vsctl = '/usr/bin/ovs-vsctl'
 ofctl = '/usr/bin/ovs-ofctl'
 cacert_filename = '/etc/openvswitch/vswitchd.cacert'
-ovsdb_port = '6640'
+ovsdb_port = '6632'
 
 
 # Delete the CA certificate, so that we go back to boot-strapping mode
